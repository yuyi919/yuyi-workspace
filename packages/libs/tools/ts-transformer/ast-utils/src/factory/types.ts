import { IMinimatch } from "minimatch";
import {} from "ts-expose-internals";
import { PluginConfig } from "ts-patch";
import ts, { CompilerOptions, EmitHost, Pattern, SourceFile } from "typescript";
import tsThree from "./declarations/typescript3";
import { HarmonyFactory } from "./utils/harmony-factory";

/* ****************************************************************************************************************** */
// region: TS Types
/* ****************************************************************************************************************** */

export type TypeScriptLatest = typeof ts;
export type TypeScriptThree = typeof tsThree;
export type ImportOrExportDeclaration = ts.ImportDeclaration | ts.ExportDeclaration;
export type ImportOrExportClause =
  | ts.ImportDeclaration["importClause"]
  | ts.ExportDeclaration["exportClause"];

// endregion

/* ****************************************************************************************************************** */
// region: Config
/* ****************************************************************************************************************** */

export interface CustomTransformerConfig extends PluginConfig {
  // readonly useRootDirs?: boolean;
  readonly exclude?: string[];
}

// endregion

/* ****************************************************************************************************************** */
// region: Contexts
/* ****************************************************************************************************************** */

export interface CustomTransformerContext<
  Config extends CustomTransformerConfig = CustomTransformerConfig
> {
  /**
   * TS Instance passed from ts-patch / ttypescript with TS4+ typings
   */
  readonly tsInstance: TypeScriptLatest;
  /**
   * TS Instance passed from ts-patch / ttypescript with TS3 typings
   */
  readonly tsThreeInstance: TypeScriptThree;
  readonly tsFactory?: ts.NodeFactory;
  readonly program?: ts.Program;
  readonly tsThreeProgram?: tsThree.Program;
  readonly config: Config;
  readonly compilerOptions: CompilerOptions;
  readonly elisionMap: Map<
    ts.SourceFile,
    Map<ImportOrExportDeclaration, ImportOrExportDeclaration>
  >;
  readonly transformationContext: ts.TransformationContext;
  readonly rootDirs?: string[];
  readonly excludeMatchers: IMinimatch[] | undefined;
  readonly outputFileNamesCache: Map<SourceFile, string>;
  readonly pathsPatterns: (string | Pattern)[] | undefined;
  readonly emitHost: EmitHost;
  readonly isTsNode: boolean;
}

export interface VisitorContext<Config extends CustomTransformerConfig = CustomTransformerConfig>
  extends CustomTransformerContext<Config> {
  readonly factory: HarmonyFactory;
  readonly sourceFile: ts.SourceFile;
  readonly isDeclarationFile: boolean;
  readonly originalSourceFile: ts.SourceFile;
  getVisitor(): (node: ts.Node) => ts.VisitResult<ts.Node>;
}
