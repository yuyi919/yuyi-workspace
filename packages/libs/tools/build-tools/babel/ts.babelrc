{
  "presets": [
    [
      "@babel/preset-typescript",
      {
        "allowNamespaces": false
      }
    ]
  ],
  "plugins": [
    [
      "const-enum",
      {
        "transform": "constObject"
      }
    ],
    [
      "import",
      {
        "libraryName": "lodash",
        "libraryDirectory": "",
        "camel2DashComponentName": false,
        "style": false
      }
    ]
  ]
}
