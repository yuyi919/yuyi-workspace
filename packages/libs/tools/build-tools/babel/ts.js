/**
 *
 * @param {import("@babel/core").ConfigAPI} api
 * @param {*} param1
 */
module.exports = function (api, { jsx }) {
  api.assertVersion(7);
  const plugins = [];
  switch (jsx) {
    case "inferno":
      plugins.push("babel-plugin-inferno");
      break;
    case "vue-next":
      plugins.push("@vue/babel-plugin-jsx");
      break;
  }
  return {
    presets: [
      [
        require.resolve("@babel/preset-typescript"),
        {
          allowNamespaces: true,
        },
      ],
    ],
    plugins: [
      [
        require.resolve("babel-plugin-const-enum"),
        {
          transform: "constObject",
        },
      ],
      ...plugins,
    ],
  };
};
