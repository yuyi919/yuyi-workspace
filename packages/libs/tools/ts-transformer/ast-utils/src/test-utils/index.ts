export * from "./spec"
export * from "./compile"

