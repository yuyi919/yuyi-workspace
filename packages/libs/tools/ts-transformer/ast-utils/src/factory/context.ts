import { Minimatch } from "minimatch";
import path from "path";
import { TransformerExtras } from "ts-patch";
import ts from "typescript";
import tsThree from "./declarations/typescript3";
import { CustomTransformerConfig, CustomTransformerContext, TypeScriptThree } from "./types";
import { cast } from "./utils/helper";
import { createSyntheticEmitHost, getTsNodeRegistrationProperties } from "./utils/ts-helper";

export function createTransformContext<TransformerConfig extends CustomTransformerConfig>(
  transformationContext: ts.TransformationContext,
  program?: ts.Program,
  pluginConfig?: TransformerConfig,
  transformerExtras?: Partial<TransformerExtras>,
  /**
   * Supply if manually transforming with compiler API via 'transformNodes' / 'transformModule'
   */
  manualTransformOptions?: {
    compilerOptions?: ts.CompilerOptions;
    fileNames?: string[];
  }
): CustomTransformerContext<TransformerConfig> {
  let tsInstance: typeof ts;
  let compilerOptions: ts.CompilerOptions;
  let fileNames: readonly string[] | undefined;
  let isTsNode = false;

  tsInstance = transformerExtras?.ts ?? ts;
  compilerOptions = manualTransformOptions?.compilerOptions!;

  if (program) {
    compilerOptions ??= program.getCompilerOptions();
  } else if (manualTransformOptions) {
    fileNames = manualTransformOptions.fileNames;
  } else {
    const tsNodeProps = getTsNodeRegistrationProperties(tsInstance);
    if (!tsNodeProps)
      throw new Error(
        `Cannot transform without a Program, ts-node instance, or manual parameters supplied. ` +
          `Make sure you're using ts-patch or ts-node with transpileOnly.`
      );
    isTsNode = true;
    compilerOptions = tsNodeProps?.compilerOptions || transformationContext.getCompilerOptions();
    fileNames = tsNodeProps?.fileNames;
  }

  const rootDirs = compilerOptions.rootDirs?.filter(path.isAbsolute);
  const config: TransformerConfig = pluginConfig ?? ({} as TransformerConfig);
  const getCanonicalFileName = tsInstance.createGetCanonicalFileName(
    tsInstance.sys.useCaseSensitiveFileNames
  );

  let emitHost = transformationContext.getEmitHost();
  if (!emitHost) {
    if (!fileNames)
      throw new Error(
        `No EmitHost found and could not determine files to be processed. Please file an issue with a reproduction!`
      );
    emitHost = createSyntheticEmitHost(
      compilerOptions,
      tsInstance,
      getCanonicalFileName,
      fileNames as string[]
    );
  } else if (isTsNode) {
    Object.assign(emitHost, { getCompilerOptions: () => compilerOptions });
  }

  const { configFile, paths } = compilerOptions;
  // TODO - Remove typecast when tryParsePatterns is recognized (probably after ts v4.4)
  const { tryParsePatterns } = tsInstance as any;
  return {
    compilerOptions,
    config,
    elisionMap: new Map(),
    tsFactory: transformationContext.factory,
    program,
    tsThreeProgram: cast<tsThree.Program>(program),
    rootDirs,
    transformationContext,
    tsInstance,
    emitHost,
    isTsNode,
    tsThreeInstance: cast<TypeScriptThree>(tsInstance),
    excludeMatchers: config.exclude?.map(
      (globPattern) => new Minimatch(globPattern, { matchBase: true })
    ),
    outputFileNamesCache: new Map(),
    // Get paths patterns appropriate for TS compiler version
    pathsPatterns:
      paths &&
      (tryParsePatterns
        ? // TODO - Remove typecast when pathPatterns is recognized (probably after ts v4.4)
          (configFile?.configFileSpecs as any)?.pathPatterns || tryParsePatterns(paths)
        : tsInstance.getOwnKeys(paths)),
  };
}
