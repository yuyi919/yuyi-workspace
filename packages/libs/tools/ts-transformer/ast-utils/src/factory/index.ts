import ts from "typescript";
import { createTransformContext } from "./context";
import { CustomTransformerConfig, VisitorContext } from "./types";
import { createHarmonyFactory } from "./utils/harmony-factory";

export type CustomTransformerHook<Config extends CustomTransformerConfig> = (
  this: VisitorContext<Config>,
  node: ts.Node,
  context: VisitorContext<Config>
) => ts.VisitResult<ts.Node>;

export function createCustomTransformer<Config extends CustomTransformerConfig>(
  nodeVisitor: CustomTransformerHook<Config>
) {
  return (transformationContext: ts.TransformationContext) => {
    const tsTransformPathsContext = createTransformContext<Config>(transformationContext);
    const { tsInstance } = tsTransformPathsContext;
    return (sourceFile: ts.SourceFile) => {
      const visitorContext: VisitorContext = {
        ...tsTransformPathsContext,
        sourceFile,
        isDeclarationFile: sourceFile.isDeclarationFile,
        originalSourceFile: (<typeof ts>tsInstance).getOriginalSourceFile(sourceFile),
        getVisitor() {
          return (node) => nodeVisitor.call(this, node, visitorContext);
        },
        factory: createHarmonyFactory(tsTransformPathsContext),
      };
      return tsInstance.visitEachChild(
        sourceFile,
        visitorContext.getVisitor(),
        transformationContext
      );
    };
  };
}
export * from "./context";
export * from "./types";
