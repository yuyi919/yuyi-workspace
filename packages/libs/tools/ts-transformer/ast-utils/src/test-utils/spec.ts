import ts from "typescript";
import { readFileSync } from "fs-extra";
import { createTransformContext } from "../factory/context";
import { CustomTransformerContext } from "../factory/types";

export type Spec = (
  node: ts.Node,
  context: CustomTransformerContext
) => ts.Node | void | undefined | false;
export type SpecFactory = (
  context: CustomTransformerContext
) => (node: ts.Node) => ts.Node | void | undefined | false;

export function spec(spec: Spec, path: string) {
  return testCode(spec, readFileSync(path).toString(), path);
}
// endregion

/* ****************************************************************************************************************** */
// region: Transformer
/* ****************************************************************************************************************** */

export function createTransformer(spec: Spec) {}

export function testCode(spec: Spec, code: string, path = "") {
  const sourceFile = ts.createSourceFile(
    path,
    code,
    ts.ScriptTarget.Latest,
    true,
    ts.ScriptKind.TS
  );
  ts.transform(sourceFile, [
    (transformationContext) => {
      const context = createTransformContext(transformationContext, void 0, {}, void 0, {
        fileNames: [path],
        compilerOptions: transformationContext.getCompilerOptions(),
      });
      return (sourceFile) => {
        function visitor(node: ts.Node) {
          const result = spec(node, context);
          if (result) {
            return result;
          }
          return ts.visitEachChild(node, visitor, transformationContext);
        }
        return ts.visitNode(sourceFile, visitor);
      };
    },
  ]);
}
