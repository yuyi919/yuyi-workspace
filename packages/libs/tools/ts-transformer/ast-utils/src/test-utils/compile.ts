import fs from "fs-extra";
import { join, relative, normalize } from "path";
import ts, {
  CompilerOptions,
  createProgram,
  CustomTransformers,
  ModuleKind,
  ScriptTarget,
  WriteFileCallback,
} from "typescript";
import { createTransformContext, CustomTransformerConfig, CustomTransformerHook, VisitorContext } from "../factory";
import { createHarmonyFactory } from "../factory/utils/harmony-factory";
import { Spec, SpecFactory } from "./spec";
export type CustomCompilerOptions<Config> = {
  throwError?: boolean;
  compilerOptions?: Partial<CompilerOptions>;
  tsConfig?: string;
  ts?: string;
  config?: Config;
};

export function transform<Config>(
  filePath: string,
  transformers: (program: ts.Program, options: CustomCompilerOptions<Config>) => CustomTransformers,
  options: CustomCompilerOptions<Config> = {}
) {
  const result: [string, string][] = [];
  compile(
    [filePath],
    transformers,
    (fileName, fileText) => {
      result.push([fileName, fileText]);
    },
    options
  );
  return result;
}
export function transformSpec<Config extends CustomTransformerConfig>(
  nodeVisitor: CustomTransformerHook<Config>,
  filePath: string,
  options?: CustomCompilerOptions<Config>
) {
  return transform(
    filePath,
    (program) => {
      const transformer: ts.TransformerFactory<ts.SourceFile> = (transformationContext) => {
        const context = createTransformContext<Config>(transformationContext, program);
        return (sourceFile) => {
          const visitorContext: VisitorContext<Config> = {
            ...context,
            sourceFile,
            isDeclarationFile: sourceFile.isDeclarationFile,
            originalSourceFile: (<typeof ts>context.tsInstance).getOriginalSourceFile(sourceFile),
            getVisitor() {
              function visitor(node: ts.Node) {
                const result = nodeVisitor.call(visitorContext, node, visitorContext);
                if (result !== void 0) {
                  return result;
                }
                return ts.visitEachChild(node, visitor, transformationContext);
              }
              return visitor;
            },
            factory: createHarmonyFactory(context),
          };
          return context.tsInstance.visitNode(sourceFile, visitorContext.getVisitor());
        };
      };
      const before = options?.config?.before ?? true
      const after = options?.config?.after || false
      const afterDeclarations = options?.config?.afterDeclarations || false
      console.log(`\nbefore %s\nafter: %s\nafterDeclarations: %s`, before, after, afterDeclarations)
      return {
        before: before && [transformer],
        after: after && [transformer],
        afterDeclarations: afterDeclarations && [transformer],
        // afterDeclarations: [transformer],
      };
    },
    options
  ).map(([path, code]) => [normalize(relative(join(__dirname, ".."), path)), code]);
}

export function compile<Config>(
  filePaths: string[],
  transformers: (program: ts.Program, options: CustomCompilerOptions<Config>) => CustomTransformers,
  writeFileCallback?: WriteFileCallback,
  {
    ts = "typescript",
    compilerOptions = {},
    tsConfig: configPath = "",
    ...options
  }: CustomCompilerOptions<Config> = {}
) {
  if (fs.existsSync(configPath)) {
    try {
      const json = fs.readJSONSync(configPath);
      compilerOptions = Object.assign({}, json.compilerOptions || {}, compilerOptions);
    } catch (error) {}
  }
  const program = (require(ts).createProgram as typeof createProgram)(filePaths, {
    strict: true,
    noEmitOnError: false,
    suppressImplicitAnyIndexErrors: true,
    target: ScriptTarget.ES5,
    esModuleInterop: true,
    module: ModuleKind.CommonJS,
    declaration: true,
    noEmitHelpers: false,
    noUnusedLocals: false,
    importHelpers: true,
    emitDecoratorMetadata: true,
    experimentalDecorators: true,
    ...compilerOptions,
  });
  const r = program.emit(
    undefined,
    writeFileCallback,
    undefined,
    false,
    transformers(program, options)
  );
  if ((!options || options.throwError) && r.emitSkipped) {
    const { diagnostics } = r;
    throw new Error(diagnostics.map((diagnostic) => diagnostic.messageText).join("\n"));
  }
  return r;
}
