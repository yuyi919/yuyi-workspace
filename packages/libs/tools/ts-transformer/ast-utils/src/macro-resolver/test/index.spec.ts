import ts from "typescript";
import { createMacroFunctionResolver } from "..";
import { CustomTransformerHook } from "../../factory";
import { transformSpec } from "../../test-utils";

describe("test", () => {
  const transformer: CustomTransformerHook<any> = (node, context) => {
    const resolver = createMacroFunctionResolver(
      require.resolve("./importSource"),
      "CREATE",
      context
    );
    if (resolver.isImportOrExport(node)) {
      return resolver.cleanImportOrExport(node);
    }
    if (resolver.isCallExpression(node)) {
      const name = node.getText();
      console.log(name, "isRefedCallExpression");
      return ts.factory.createObjectLiteralExpression([]);
    }
  };
  it("具名导入", () => {
    const result = transformSpec(transformer, require.resolve("./named"), {
      config: {
        // afterDeclarations: true,
      },
      compilerOptions: {
        target: ts.ScriptTarget.ESNext,
        module: ts.ModuleKind.ESNext,
      },
    });
    expect(result).toMatchSnapshot("named");
  });
  it("命名空间导入", () => {
    const result = transformSpec(transformer, require.resolve("./namespaced"), {
      config: {
        // afterDeclarations: true,
      },
      compilerOptions: {
        target: ts.ScriptTarget.ESNext,
        module: ts.ModuleKind.ESNext,
      },
    });
    expect(result).toMatchSnapshot();
  });
});
