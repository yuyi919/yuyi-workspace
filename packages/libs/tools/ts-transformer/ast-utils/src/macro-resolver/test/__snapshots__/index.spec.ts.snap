// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 具名导入: named 1`] = `
Array [
  Array [
    "macro-resolver\\\\test\\\\importSource.js",
    "export function CREATE() {
    return {};
}
export const result = 1;
",
  ],
  Array [
    "macro-resolver\\\\test\\\\importSource.d.ts",
    "export declare function CREATE(): {};
export declare const result = 1;
",
  ],
  Array [
    "macro-resolver\\\\test\\\\named.js",
    "import { result } from \\"./importSource\\";
export const a = {};
export { result };
",
  ],
  Array [
    "macro-resolver\\\\test\\\\named.d.ts",
    "import { result } from \\"./importSource\\";
export declare const a: {};
export { result };
",
  ],
]
`;

exports[`test 命名空间导入 1`] = `
Array [
  Array [
    "macro-resolver\\\\test\\\\importSource.js",
    "export function CREATE() {
    return {};
}
export const result = 1;
",
  ],
  Array [
    "macro-resolver\\\\test\\\\importSource.d.ts",
    "export declare function CREATE(): {};
export declare const result = 1;
",
  ],
  Array [
    "macro-resolver\\\\test\\\\namespaced.js",
    "import * as T from \\"./importSource\\";
export const a = {};
",
  ],
  Array [
    "macro-resolver\\\\test\\\\namespaced.d.ts",
    "export declare const a: {};
",
  ],
]
`;
